### PR TITLE
feat: Vault Sunsetting functionality

### PR DESCRIPTION
**Pull request by mrod502**
Changes:
- Open up permissions on vault `nextEpoch` function so that a user with deposited or pending deposit funds (token0 or token1) can trigger.
- Add a storage variable `depositsEnabled` and associated setter/modifier for `depositToken0/1` so we can disable deposits.

Will merge these changes to `disable-deposits` branch to preserve original vault functionality in `master`

### CODE REVIEW

File: src/test/utils/VaultUtils.sol
Comment (ttarsi) [inline src/test/utils/VaultUtils.sol]: let's use `amount` here, which is a storage variable defined above. E.g. we don't need this line and can just use `amount` whenever we used `_amount` here

File: src/vaults/VaultStorage.sol
Comment (ttarsi) [inline src/vaults/VaultStorage.sol]: This isn't necessary – since the next storage variable is a uint256 it will [automatically use the next storage slot](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#layout-of-state-variables-in-storage). But this line can be helpful as a reminder that we can save space if we need to use other non-uint256 storage vars in the future.

File: src/vaults/Vault.sol
Comment (ttarsi) [inline src/vaults/Vault.sol]: Let's add a `setDepositsDisabled` function as well

File: src/vaults/Vault.sol
Comment (ttarsi) [inline src/vaults/Vault.sol:759]: Let's put modifiers before function defs: https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#layout-of-state-variables-in-storage

**Review by ttarsi — COMMENTED** [review]
Nice - looks like this achieves what we want it to. I'm proposing a couple things to make the upgrade process smoother:
- We place this in its own branch, called `disable-deposits`. `master` can remain as normal vault functionality if anyone (or ourselves) wants to use them in the future
- We set the `depositsEnabled` variable to false by default. When we upgrade the contracts, deposits will be automatically disabled. Thus we can minimize the number of multisig transactions required

File: src/vaults/VaultStorage.sol
Comment (mrod502) [reply src/vaults/VaultStorage.sol]: Will remove since we likely won't be making storage updates in the future, so this won't serve much purpose

File: src/vaults/Vault.sol
Comment (mrod502) [reply src/vaults/Vault.sol:759]: I see this is the pattern in our files. Does this have an impact on storage layout? I don't see anything in the included link related to modifier/function ordering.

Comment (ttarsi) [reply src/vaults/Vault.sol:759]: Sorry, I used the wrong link - [this is the correct one](https://docs.soliditylang.org/en/v0.8.16/style-guide.html#order-of-layout). No impact, just a style guide

File: src/test/utils/VaultUtils.sol
Comment (ttarsi) [inline src/test/utils/VaultUtils.sol:42]: 👍

File: src/test/utils/VaultUtils.sol
Comment (ttarsi) [inline src/test/utils/VaultUtils.sol]: we can remove this since `vm.prank()` only pranks the next call. `vm.stopPrank` is relevant when we use `vm.startPrank` because then it applies to all the external calls until `stopPrank` is used

File: src/test/utils/VaultUtils.sol
Comment (ttarsi) [inline src/test/utils/VaultUtils.sol]: is this required because the vault fails to rollover if there are no deposits at all, instead of failing with `NOT_PARTICIPANT_OR_STRATEGIST`?

File: src/test/utils/VaultUtils.sol
Comment (ttarsi) [inline src/test/utils/VaultUtils.sol]: hm.. this shouldn't be necessary unless we called `startPrank` without calling `stopPrank` previously, but we shouldn't really do that

**Review by ttarsi — APPROVED** [review]
Ok great - small nits on some forge details but lgtm. Let's wait for a 2nd review from @kevinhalliday as well

File: src/test/utils/VaultUtils.sol
Comment (mrod502) [reply src/test/utils/VaultUtils.sol]: It seems like this isn't necessary, as the revert will happen before any vault logic takes place. Removing.

File: src/test/utils/VaultUtils.sol
Comment (mrod502) [reply src/test/utils/VaultUtils.sol]: When debugging a failing test, there was a message along the lines of "there is already an active prank", so I added this. It seems to work fine without this now though.

File: src/vaults/VaultStorage.sol
Comment (kevinhalliday) [inline src/vaults/VaultStorage.sol]: the 100 words of storage comment looks weird now. but we can't just change both to 99. maybe we add some comments explaining which storage was added in this upgrade

File: src/test/utils/VaultUtils.sol
Comment (kevinhalliday) [inline src/test/utils/VaultUtils.sol]: let's throw in a `advanceWithDepositedToken0` test case too, just to be safe, because that's what most users have

**Review by kevinhalliday — COMMENTED** [review]
Looks good! Added two comments for discussion

**Review by kevinhalliday — APPROVED** [review]
👏

**Pull request merged**
